File: 1428-leftmost-column-with-at-least-a-one/1428-leftmost-column-with-at-least-a-one.py
# """
# This is BinaryMatrix's API interface.
# You should not implement it, or speculate about its implementation
# """
# class BinaryMatrix(object):
#    def get(self, row: int, col: int) -> int:
#    def dimensions(self) -> list[]:


class Solution:
    MAX_MATRIX_LENGTH = 100

    def leftMostColumnWithOne(self, binaryMatrix) -> int:

        self.binaryMatrix = binaryMatrix
        self.rows, self.cols = binaryMatrix.dimensions()
        self.leftmost_column_with_one = self.MAX_MATRIX_LENGTH + 1

        rows_with_ones = self.get_rows_with_ones()

        for row in rows_with_ones:
            self.binary_search_row(row)

        return (
            self.leftmost_column_with_one
            if self.leftmost_column_with_one <= self.MAX_MATRIX_LENGTH
            else -1
        )

    def binary_search_row(self, row: int) -> None:
        left, right = 0, min(self.leftmost_column_with_one, self.cols - 1)
        if self.binaryMatrix.get(row, right) == 0:
            return

        while left <= right:
            mid = (left + right) // 2
            if self.binaryMatrix.get(row, mid) == 1:
                self.leftmost_column_with_one = min(
                    self.leftmost_column_with_one, mid
                )
                right = mid - 1
            else:
                left = mid + 1
                if left >= self.leftmost_column_with_one:
                    break

    def get_rows_with_ones(self) -> set:
        rows_with_ones = set()
        for row in range(self.rows):
            if self.binaryMatrix.get(row, self.cols - 1) == 1:
                rows_with_ones.add(row)
        return rows_with_ones

File: 1428-leftmost-column-with-at-least-a-one/1428-leftmost-column-with-at-least-a-one.cs
class Solution
    {
    private const int MaxMatrixLength = 100;
    private BinaryMatrix _binaryMatrix;
    private int _rows;
    private int _cols;
    private int _leftmostColumnWithOne = MaxMatrixLength + 1;

    public int LeftMostColumnWithOne(BinaryMatrix binaryMatrixParam)
    {
        _binaryMatrix = binaryMatrixParam;
        _rows = _binaryMatrix.Dimensions()[0];
        _cols = _binaryMatrix.Dimensions()[1];

        var rowsWithOnes = GetRowsWithOnes();

        foreach (var row in rowsWithOnes) BinarySearchRow(row);

        if (_leftmostColumnWithOne == MaxMatrixLength + 1)
            return -1;
        return _leftmostColumnWithOne;
    }

    private HashSet<int> GetRowsWithOnes()
    {
        var rowsWithOnes = new HashSet<int>();
        for (var row = 0; row < _rows; row++)
            if (_binaryMatrix.Get(row, _cols - 1) == 1)
                rowsWithOnes.Add(row);
        return rowsWithOnes;
    }

    private void BinarySearchRow(int row)
    {
        var left = 0;
        var right = _cols - 1;

        while (left <= right)
        {
            var mid = (left + right) / 2;
            if (_binaryMatrix.Get(row, mid) == 1)
            {
                _leftmostColumnWithOne = Math.Min(_leftmostColumnWithOne, mid);
                right = mid - 1;
            }
            else
            {
                left = mid + 1;
                if (left >= _leftmostColumnWithOne)
                    break;
            }
        }
    }
}